### ISSUE TITLE
Overlapping boundary attributes not inherited in submesh

### ISSUE THREAD

**Issue by Heinrich-BR** [mfem/mfem#4387]
Hi MFEM devs,

there are meshes that I use where certain faces were assigned to more than one sideset, i.e. they have more than one boundary attribute. This is generally handled fine in MFEM when importing the mesh as a `ParMesh`. The `ParMesh` object maintains the overlapping boundary attributes. However, if I then extract a `ParSubMesh` out of it, the faces which have two (or more) boundary attributes will only inherit one of them. I was wondering if it would be possible to have the `ParSubMesh` object inherit the boundary attributes from its parent fully, including overlapping attributes.

**Comment by v-dobrev** [comment]
@Heinrich-BR, thanks for reporting this issue!

@jandrej, @mlstowell, this sounds vaguely familiar -- do we have another issues open regarding this?

**Comment by Heinrich-BR** [comment]
Hi @v-dobrev, thank you for your response! This issue is somewhat similar to one that I opened myself a while ago (#4011), although not the same. Perhaps that's the one you're thinking about?

**Comment by v-dobrev** [comment]
> Hi @v-dobrev, thank you for your response! This issue is somewhat similar to one that I opened myself a while ago (#4011), although not the same. Perhaps that's the one you're thinking about?

Okay, yeah, you're probably right. The important point is that this issue is different from #4011, so we need to investigate it.

**Comment by jandrej** [comment]
https://github.com/mfem/mfem/blob/9343ffad7b64de9ccf29ad23c740bb02d916c3e1/mesh/submesh/submesh.cpp#L71-L72
We inherit the elements boundary attribute that belongs to the submesh. How can an element have multiple attributes?

**Comment by v-dobrev** [comment]
I think the new mesh sets that Mark added allow you to put the same element/bdr element/face/edge/vertex in different mesh sets. At least that's what I assumed this is about.

**Comment by Heinrich-BR** [comment]
Hi @v-dobrev and @jandrej! Thank you for your comments. Let me be more specific about hat I mean with an example. I've created on Coreform Cubit a 3D mesh which is just a parallelepiped of dimensions 5x5x10, like so:

![Screenshot from 2024-07-17 10-28-32](https://github.com/user-attachments/assets/40a4d934-4509-4755-8a34-fc13db3ad1c2)
 
There are 3 sets of boundary attributes that I've created for this:

Boundary attribute 1 includes every surface in in the mesh.

Boundary attribute 2 includes one side and the two "lids", like so:

![Screenshot from 2024-07-17 10-28-38](https://github.com/user-attachments/assets/ec56c71b-1633-493f-a6e0-aee919e7e4ee)

Boundary attribute 2 includes another side and the two "lids" again, like so:

![Screenshot from 2024-07-17 10-28-45](https://github.com/user-attachments/assets/90c20e11-b208-4207-922c-7f88c7a9715b)

So what we can see from this is that in this mesh we have examples of sides which are included in one, two, or even three sidesets. The two lids are included in 3 sidesets. Two of the sides are included in two sidesets, and another two sides are included in one sideset. I'll be happy to send this mesh file over if you'd like.

Now consider running this minimal example in the most current version of MFEM as of this comment:

```   
   // Import and save the mesh
   string mesh_file = "sideset_example.e";
   Mesh mesh(mesh_file);
   mesh.Save("sideset_after.mesh");

   // Now create and save a submesh
   Array<int> attr;
   attr.Append(1);
   SubMesh submesh(SubMesh::CreateFromDomain(mesh, attr));
   submesh.Save("submesh_after.mesh");
```

We import this mesh as an MFEM `Mesh` object and then immediately save it as "sideset_after.mesh". If you now go on GLVis, you'll see that the sidesets are unaltered, and still overlap as they should. This points to the fact that the `Mesh` object does support overlapping sidesets, i.e. faces with more than one boundary attribute. However, if we now have a look at at the submesh that was extracted, what we get is as follows:

Boundary attribute 1:

![Screenshot from 2024-07-17 10-46-25](https://github.com/user-attachments/assets/ffad8249-3f9a-47dc-9b03-c72f862871ec)

Boundary attribute 2:

![Screenshot from 2024-07-17 10-46-31](https://github.com/user-attachments/assets/0abeb97d-60f7-4d3b-8505-1c4512bed939)

Boundary attribute 3:

![Screenshot from 2024-07-17 10-46-35](https://github.com/user-attachments/assets/8be39354-f151-40d0-8eb5-c43c9afe2da3)

We can see that now the boundary attributes were not inherited as they were on the original mesh. There is no overlap between them. Rather, any particular face inherits only the last boundary attribute that was assigned to it. Hopefully what I mean now is clearer. While my previous issue with submesh face inheritance that was raised in #4011 has been fixed since, this particular issue remains. Let me know if I can help in any way or if you'd like me to send you this example mesh I've used. Hope you have a great day!

**Comment by jandrej** [comment]
Thanks for your detailed description. Is Boundary attribute 3 of the original mesh the last overwritten state s.t. submesh takes the last state in the figure below?
![image](https://github.com/user-attachments/assets/338f4f95-4cc5-4dda-af55-e1344246f83e)

**Comment by Heinrich-BR** [comment]
Yes, indeed. In the extracted submesh, boundary attribute 3 is the only one that is being inherited as you would expect (it's the upper half of boundary attribute 3 on the original mesh). I believe this is because it's the last boundary attribute to be set, so that it overrides the others, as opposed to being added to them.

**Comment by v-dobrev** [comment]
@Heinrich-BR, can you post the mesh file `sideset_example.e` here?

**Comment by Heinrich-BR** [comment]
Of course! I've zipped it since git doesn't support `.e` files. Here it is: [example_sideset.zip](https://github.com/user-attachments/files/16281426/example_sideset.zip)

**Comment by Heinrich-BR** [comment]
Hi @v-dobrev and @jandrej, just to follow up, have you managed to examine the example mesh file? Let me know if there are any issues
